In [212]:
import numpy as np
import pandas as pd

In [213]:
df = pd.read_csv('Bengaluru_House_Data.csv')

In [214]:
df.head()

area_type   availability                  location       size  \
0  Super built-up  Area         19-Dec  Electronic City Phase II      2 BHK   
1            Plot  Area  Ready To Move          Chikka Tirupathi  4 Bedroom   
2        Built-up  Area  Ready To Move               Uttarahalli      3 BHK   
3  Super built-up  Area  Ready To Move        Lingadheeranahalli      3 BHK   
4  Super built-up  Area  Ready To Move                  Kothanur      2 BHK   

   society total_sqft  bath  balcony   price  
0  Coomee        1056   2.0      1.0   39.07  
1  Theanmp       2600   5.0      3.0  120.00  
2      NaN       1440   2.0      3.0   62.00  
3  Soiewre       1521   3.0      1.0   95.00  
4      NaN       1200   2.0      1.0   51.00

In [215]:
df.shape

(13320, 9)

In [216]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13320 entries, 0 to 13319
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   area_type     13320 non-null  object 
 1   availability  13320 non-null  object 
 2   location      13319 non-null  object 
 3   size          13304 non-null  object 
 4   society       7818 non-null   object 
 5   total_sqft    13320 non-null  object 
 6   bath          13247 non-null  float64
 7   balcony       12711 non-null  float64
 8   price         13320 non-null  float64
dtypes: float64(3), object(6)
memory usage: 936.7+ KB


In [217]:
for column in df.columns:
    print(df[column].value_counts())
    print('*'*46)

Super built-up  Area    8790
Built-up  Area          2418
Plot  Area              2025
Carpet  Area              87
Name: area_type, dtype: int64
**********************************************
Ready To Move    10581
18-Dec             307
18-May             295
18-Apr             271
18-Aug             200
                 ...  
15-Dec               1
14-Jul               1
17-Jan               1
16-Jul               1
15-Aug               1
Name: availability, Length: 81, dtype: int64
**********************************************
Whitefield                    540
Sarjapur  Road                399
Electronic City               302
Kanakpura Road                273
Thanisandra                   234
                             ... 
B K Nagar                       1
Mukkutam Nagar                  1
Yemlur, Old Airport Road,       1
Kuvempu Layout                  1
Sunkan palya                    1
Name: location, Length: 1305, dtype: int64
*********************************************

In [218]:
df.isna().sum()

area_type          0
availability       0
location           1
size              16
society         5502
total_sqft         0
bath              73
balcony          609
price              0
dtype: int64

# as 'society' column has many blank values so it's not at all useful

In [219]:
df.drop(columns=['society','availability','balcony','area_type'],inplace=True)

In [220]:
df.describe()

bath         price
count  13247.000000  13320.000000
mean       2.692610    112.565627
std        1.341458    148.971674
min        1.000000      8.000000
25%        2.000000     50.000000
50%        2.000000     72.000000
75%        3.000000    120.000000
max       40.000000   3600.000000

In [221]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13320 entries, 0 to 13319
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   location    13319 non-null  object 
 1   size        13304 non-null  object 
 2   total_sqft  13320 non-null  object 
 3   bath        13247 non-null  float64
 4   price       13320 non-null  float64
dtypes: float64(2), object(3)
memory usage: 520.4+ KB


In [222]:
df['location'].value_counts()

Whitefield                    540
Sarjapur  Road                399
Electronic City               302
Kanakpura Road                273
Thanisandra                   234
                             ... 
B K Nagar                       1
Mukkutam Nagar                  1
Yemlur, Old Airport Road,       1
Kuvempu Layout                  1
Sunkan palya                    1
Name: location, Length: 1305, dtype: int64

In [223]:
df['location'] = df['location'].fillna('Sarjapur Road')

In [224]:
df['size'].value_counts()

2 BHK         5199
3 BHK         4310
4 Bedroom      826
4 BHK          591
3 Bedroom      547
1 BHK          538
2 Bedroom      329
5 Bedroom      297
6 Bedroom      191
1 Bedroom      105
8 Bedroom       84
7 Bedroom       83
5 BHK           59
9 Bedroom       46
6 BHK           30
7 BHK           17
1 RK            13
10 Bedroom      12
9 BHK            8
8 BHK            5
10 BHK           2
11 BHK           2
11 Bedroom       2
43 Bedroom       1
27 BHK           1
18 Bedroom       1
13 BHK           1
12 Bedroom       1
19 BHK           1
16 BHK           1
14 BHK           1
Name: size, dtype: int64

In [225]:
df['size'] = df['size'].fillna('2 BHK')

In [226]:
df['bath'] = df['bath'].fillna(df['bath'].median())

In [227]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13320 entries, 0 to 13319
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   location    13320 non-null  object 
 1   size        13320 non-null  object 
 2   total_sqft  13320 non-null  object 
 3   bath        13320 non-null  float64
 4   price       13320 non-null  float64
dtypes: float64(2), object(3)
memory usage: 520.4+ KB


# Now we don't have any NA in database

In [228]:
df['bhk'] = df['size'].str.split().str.get(0).astype(int)

In [229]:
df[df.bhk>20]           # Outliers --> have to removed

location        size total_sqft  bath  price  bhk
1718  2Electronic City Phase II      27 BHK       8000  27.0  230.0   27
4684                Munnekollal  43 Bedroom       2400  40.0  660.0   43

In [230]:
df['total_sqft'].unique()

array(['1056', '2600', '1440', ..., '1133 - 1384', '774', '4689'],
      dtype=object)

In [231]:
def convertRange(x):
    t = x.split('-')
    if len(t)==2:
        return (float(t[0])+float(t[1]))/2
    try:
        return float(x)
    except:
        return None

In [232]:
df['total_sqft'] = df['total_sqft'].apply(convertRange)

In [233]:
df.head()

location       size  total_sqft  bath   price  bhk
0  Electronic City Phase II      2 BHK      1056.0   2.0   39.07    2
1          Chikka Tirupathi  4 Bedroom      2600.0   5.0  120.00    4
2               Uttarahalli      3 BHK      1440.0   2.0   62.00    3
3        Lingadheeranahalli      3 BHK      1521.0   3.0   95.00    3
4                  Kothanur      2 BHK      1200.0   2.0   51.00    2

In [234]:
df['total_sqft'].value_counts()

1200.00    843
1100.00    221
1500.00    206
2400.00    196
600.00     180
          ... 
2489.00      1
2112.95      1
3640.00      1
3385.50      1
2791.00      1
Name: total_sqft, Length: 1999, dtype: int64

# Price per Square ft

In [235]:
df['price_per_sqft'] = df['price']*100000 / df['total_sqft']

In [236]:
df['price_per_sqft']

0         3699.810606
1         4615.384615
2         4305.555556
3         6245.890861
4         4250.000000
             ...     
13315     6689.834926
13316    11111.111111
13317     5258.545136
13318    10407.336319
13319     3090.909091
Name: price_per_sqft, Length: 13320, dtype: float64

In [237]:
df.describe()

total_sqft          bath         price           bhk  price_per_sqft
count  13274.000000  13320.000000  13320.000000  13320.000000    1.327400e+04
mean    1559.626694      2.688814    112.565627      2.802778    7.907501e+03
std     1238.405258      1.338754    148.971674      1.294496    1.064296e+05
min        1.000000      1.000000      8.000000      1.000000    2.678298e+02
25%     1100.000000      2.000000     50.000000      2.000000    4.266865e+03
50%     1276.000000      2.000000     72.000000      3.000000    5.434306e+03
75%     1680.000000      3.000000    120.000000      3.000000    7.311746e+03
max    52272.000000     40.000000   3600.000000     43.000000    1.200000e+07

In [238]:
df['location'].value_counts()

Whitefield                    540
Sarjapur  Road                399
Electronic City               302
Kanakpura Road                273
Thanisandra                   234
                             ... 
Mukkutam Nagar                  1
Yemlur, Old Airport Road,       1
Kuvempu Layout                  1
Sadduguntepalya                 1
Sunkan palya                    1
Name: location, Length: 1306, dtype: int64

# here 1306 locations are there and if we use onehotencoding on it then we will get 1306 extra columns to tackle this we'll convert the locations which have occured less than 10 times to 'other'

In [239]:
df['location'] = df['location'].apply(lambda x: x.strip())

In [240]:
location_count = df['location'].value_counts()

In [241]:
location_count

Whitefield                            541
Sarjapur  Road                        399
Electronic City                       304
Kanakpura Road                        273
Thanisandra                           237
                                     ... 
Chikkaballapur                          1
Hommadevanahalli                        1
Ganganahalli                            1
Kuvempu Layout                          1
Kengeri Satellite Town ( BDA SITE)      1
Name: location, Length: 1295, dtype: int64

In [242]:
loc_count_less_than_10 = location_count[location_count<=10]
loc_count_less_than_10

Nagadevanahalli                       10
Sadashiva Nagar                       10
Ganga Nagar                           10
Sector 1 HSR Layout                   10
1st Block Koramangala                 10
                                      ..
Chikkaballapur                         1
Hommadevanahalli                       1
Ganganahalli                           1
Kuvempu Layout                         1
Kengeri Satellite Town ( BDA SITE)     1
Name: location, Length: 1054, dtype: int64

In [243]:
df['location'] = df['location'].apply(lambda x: 'other' if x in loc_count_less_than_10 else x)
df['location'].value_counts()

other                 2886
Whitefield             541
Sarjapur  Road         399
Electronic City        304
Kanakpura Road         273
                      ... 
Tindlu                  11
Nehru Nagar             11
Narayanapura            11
HAL 2nd Stage           11
Pattandur Agrahara      11
Name: location, Length: 242, dtype: int64

# outlier detection and removal


In [244]:
df.describe()

total_sqft          bath         price           bhk  price_per_sqft
count  13274.000000  13320.000000  13320.000000  13320.000000    1.327400e+04
mean    1559.626694      2.688814    112.565627      2.802778    7.907501e+03
std     1238.405258      1.338754    148.971674      1.294496    1.064296e+05
min        1.000000      1.000000      8.000000      1.000000    2.678298e+02
25%     1100.000000      2.000000     50.000000      2.000000    4.266865e+03
50%     1276.000000      2.000000     72.000000      3.000000    5.434306e+03
75%     1680.000000      3.000000    120.000000      3.000000    7.311746e+03
max    52272.000000     40.000000   3600.000000     43.000000    1.200000e+07

In [245]:
# there can't be 1sq flat so remove it

In [246]:
(df['total_sqft']/df['bhk']).describe()

count    13274.000000
mean       575.074878
std        388.205175
min          0.250000
25%        473.333333
50%        552.500000
75%        625.000000
max      26136.000000
dtype: float64

In [247]:
df = df[((df['total_sqft']/df['bhk'])>=300)]
df.describe()

total_sqft          bath         price           bhk  price_per_sqft
count  12530.000000  12530.000000  12530.000000  12530.000000    12530.000000
mean    1594.564544      2.559537    111.382401      2.650838     6303.979357
std     1261.271296      1.077938    152.077329      0.976678     4162.237981
min      300.000000      1.000000      8.440000      1.000000      267.829813
25%     1116.000000      2.000000     49.000000      2.000000     4210.526316
50%     1300.000000      2.000000     70.000000      3.000000     5294.117647
75%     1700.000000      3.000000    115.000000      3.000000     6916.666667
max    52272.000000     16.000000   3600.000000     16.000000   176470.588235

In [248]:
df.shape

(12530, 7)

In [249]:
df.price_per_sqft.describe()

count     12530.000000
mean       6303.979357
std        4162.237981
min         267.829813
25%        4210.526316
50%        5294.117647
75%        6916.666667
max      176470.588235
Name: price_per_sqft, dtype: float64

In [250]:
def remove_sqft_outliers(df):
    df_ret = pd.DataFrame()
    for key,subdf in df.groupby('location'):

        m = np.mean(subdf.price_per_sqft)
        st = np.std(subdf.price_per_sqft)

        gen_df = subdf[(subdf.price_per_sqft > (m-st)) & (subdf.price_per_sqft <= (m+st))]

        df_ret = pd.concat([df_ret,gen_df],ignore_index=True)
    return df_ret

In [251]:
# key = location , subdf = df corresponding to location
# taking only 1sigma values for location

In [252]:
df = remove_sqft_outliers(df)
df.describe()

total_sqft          bath         price           bhk  price_per_sqft
count  10301.000000  10301.000000  10301.000000  10301.000000    10301.000000
mean    1508.440608      2.471702     91.286372      2.574896     5659.062876
std      880.694214      0.979449     86.342786      0.897649     2265.774749
min      300.000000      1.000000     10.000000      1.000000     1250.000000
25%     1110.000000      2.000000     49.000000      2.000000     4244.897959
50%     1286.000000      2.000000     67.000000      2.000000     5175.600739
75%     1650.000000      3.000000    100.000000      3.000000     6428.571429
max    30400.000000     16.000000   2200.000000     16.000000    24509.803922

In [253]:
df.shape

(10301, 7)

In [256]:
def remove_bhk_outlier(df):

    indices_to_be_exclude = np.array([])
    for location, location_df in df.groupby('location'):
        bhk_stats = {}
        for bhk, bhk_df in location_df.groupby('bhk'):
            bhk_stats[bhk] = {
                'mean' : np.mean(bhk_df.price_per_sqft),
                'std'  : np.std(bhk_df.price_per_sqft),
                'count': bhk_df.shape[0]
            }
        for bhk, bhk_df in location_df.groupby('bhk'):
            stats = bhk_stats.get(bhk-1)            # if bhk-1 is not in key then get function returns none
            if stats and stats['count']>5:
                indices_to_be_exclude = np.append(indices_to_be_exclude,bhk_df[bhk_df.price_per_sqft<(stats['mean'])].index.values)
    return df.drop(indices_to_be_exclude,axis = 'index')

In [257]:
df = remove_bhk_outlier(df)

In [258]:
df.shape

(7361, 7)

In [260]:
df = df.drop(['size','price_per_sqft'],axis='columns')

In [261]:
df.to_csv('Cleaned.csv')

In [264]:
X = df.drop(columns=['price'])
y = df['price']

In [262]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression,Lasso,Ridge
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.compose import make_column_transformer
from sklearn.pipeline import make_pipeline
from sklearn.metrics import r2_score

In [265]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

In [266]:
print(X_train.shape)
print(X_test.shape)

(5888, 4)
(1473, 4)


In [267]:
column_trans = make_column_transformer((OneHotEncoder(sparse=False),['location']),remainder='passthrough')

In [268]:
scaler = StandardScaler()

In [269]:
lr = LinearRegression(normalize=True)

In [270]:
pipe = make_pipeline(column_trans,scaler,lr)

In [271]:
pipe.fit(X_train,y_train)

Pipeline(steps=[('columntransformer',
                 ColumnTransformer(remainder='passthrough',
                                   transformers=[('onehotencoder',
                                                  OneHotEncoder(sparse=False),
                                                  ['location'])])),
                ('standardscaler', StandardScaler()),
                ('linearregression', LinearRegression(normalize=True))])

In [274]:
y_train_pred_lr = pipe.predict(X_train)
r2_score(y_train,y_train_pred_lr)

0.8619356816981333

In [273]:
y_test_pred_lr = pipe.predict(X_test)
r2_score(y_test,y_test_pred_lr)

0.8233716332464034

# Applying Lasso

In [275]:
lasso = Lasso()

In [276]:
pipe = make_pipeline(column_trans,scaler, lasso)

In [277]:
pipe.fit(X_train,y_train)

Pipeline(steps=[('columntransformer',
                 ColumnTransformer(remainder='passthrough',
                                   transformers=[('onehotencoder',
                                                  OneHotEncoder(sparse=False),
                                                  ['location'])])),
                ('standardscaler', StandardScaler()), ('lasso', Lasso())])

In [280]:
y_train_pred_lasso = pipe.predict(X_train)
r2_score(y_train, y_train_pred_lasso)

0.849726035204704

In [278]:
y_test_pred_lasso = pipe.predict(X_test)
r2_score(y_test, y_test_pred_lasso)

0.8128285650772719

In [281]:
ridge = Ridge()

In [282]:
pipe = make_pipeline(column_trans,scaler,ridge)

In [283]:
pipe.fit(X_train,y_train)

Pipeline(steps=[('columntransformer',
                 ColumnTransformer(remainder='passthrough',
                                   transformers=[('onehotencoder',
                                                  OneHotEncoder(sparse=False),
                                                  ['location'])])),
                ('standardscaler', StandardScaler()), ('ridge', Ridge())])

In [284]:
y_train_pred_ridge = pipe.predict(X_train)
r2_score(y_train,y_train_pred_ridge)

0.8619601044018195

In [285]:
y_test_pred_ridge = pipe.predict(X_test)
r2_score(y_test,y_test_pred_ridge)

0.8234146633312647